File: src/agents/audio.py
from io import BytesIO

import requests
from pydub import AudioSegment  # type: ignore

from src.asset import AssetStore
from src.config import get_speakers, load_config
from src.models import Script


class AudioAgent:
    def __init__(self, store: AssetStore):
        self.store = store
        cfg = load_config()
        self.base_url = cfg["providers"]["tts"]["voicevox"]["url"]
        self.speakers = get_speakers()

    def run(self, script: Script) -> list[str]:
        self.store.log_input("audio", script.model_dump())

        audio_dir = self.store.audio_dir()
        audio_paths = []
        combined = AudioSegment.empty()

        for i, line in enumerate(script.lines):
            spk_id = self.speakers.get(line.speaker, 1)
            text = line.text

            q = requests.post(f"{self.base_url}/audio_query", params={"text": text, "speaker": spk_id})  # type: ignore
            q.raise_for_status()
            q_data = q.json()

            s = requests.post(f"{self.base_url}/synthesis", params={"speaker": spk_id}, json=q_data)  # type: ignore
            s.raise_for_status()

            chunk = AudioSegment.from_file(BytesIO(s.content), format="wav")
            combined += chunk

            path = audio_dir / f"{i:03d}_{line.speaker}.wav"
            chunk.export(str(path), format="wav")
            audio_paths.append(str(path))

        full_path = audio_dir / "full_audio.wav"
        combined.export(str(full_path), format="wav")

        self.store.log_output("audio", {"paths": audio_paths, "full": str(full_path)})

        return audio_paths

File: src/agents/audio.ts

import path from "path";
import fs from "fs-extra";
import axios from "axios";
import ffmpeg from "fluent-ffmpeg";
import { AssetStore } from "../asset.js";
import { loadConfig, getSpeakers } from "../config.js";
import { Script } from "../models.js";
import os from "os";

export class AudioAgent {
    store: AssetStore;
    baseUrl: string;
    speakers: Record<string, number>;

    constructor(store: AssetStore) {
        this.store = store;
        const cfg = loadConfig();
        this.baseUrl = cfg.providers.tts.voicevox.url;
        this.speakers = getSpeakers();
    }

    async run(script: Script): Promise<string[]> {
        this.store.logInput("audio", script);

        const audioDir = this.store.audioDir();
        const audioPaths: string[] = [];

        for (let i = 0; i < script.lines.length; i++) {
            const line = script.lines[i];
            const spkId = this.speakers[line.speaker] || 1;
            const text = line.text;

            const q = await axios.post(`${this.baseUrl}/audio_query`, null, {
                params: { text: text, speaker: spkId }
            });
            const qData = q.data;

            const s = await axios.post(`${this.baseUrl}/synthesis`, qData, {
                params: { speaker: spkId },
                responseType: 'arraybuffer'
            });

            const fileName = `${String(i).padStart(3, '0')}_${line.speaker}.wav`;
            const filePath = path.join(audioDir, fileName);
            fs.writeFileSync(filePath, Buffer.from(s.data));
            audioPaths.push(filePath);
        }

        if (audioPaths.length > 0) {
            const fullPath = path.join(audioDir, "full_audio.wav");
            await this.mergeAudio(audioPaths, fullPath);
            this.store.logOutput("audio", { paths: audioPaths, full: fullPath });
        }

        return audioPaths;
    }

    async mergeAudio(inputs: string[], output: string): Promise<void> {
        return new Promise((resolve, reject) => {
            const cmd = ffmpeg();
            inputs.forEach(i => cmd.input(i));

            cmd.on('error', (err) => {
                reject(err);
            })
                .on('end', () => {
                    resolve();
                })
                .mergeToFile(output, os.tmpdir());
        });
    }
}
